fakenet/listeners/ProxyListener.py

```diff
@@ -0,0 +1,397 @@
+                
+                while True:
+                    readable, writable, exceptional = select.select(
+                            [remote_sock], [], [], .001)
+                    if readable:
+                        try:
+                            if ssl_remote_sock:
+                                data = ssl_remote_sock.recv(BUF_SZ)
+                            else:
+                                data = remote_sock.recv(BUF_SZ)
+                            if data:
+                                remote_q.put(data)
+                            else:
+                                self.server.logger.debug(
+                                        'Closing remote socket connection')
+                                return
+                        except Exception as e:
+                            self.server.logger.debug('Remote Connection terminated')
+                            return
+                    if not listener_q.empty():
+                        data = listener_q.get()
+                        if ssl_remote_sock:
+                            ssl_remote_sock.send(data)
+                        else:
+                            remote_sock.send(data)
+
+class ThreadedUDPRequestHandler(SocketServer.BaseRequestHandler):
+
+
+    def handle(self):
+
+        data = self.request[0]
+        remote_sock = self.request[1]
+
+        self.server.logger.debug('Received UDP packet from %s.' % 
+                self.client_address[0])
+
+        if data:
+
+            self.server.logger.debug('Packet data\n%s' % hexdump.hexdump(data, 
+                result='return'))
+
+            orig_src_ip = self.client_address[0]
+            orig_src_port = self.client_address[1]
+
+            top_listener = get_top_listener(self.server.config, data, 
+                    self.server.listeners, self.server.diverter, 
+                    orig_src_ip, orig_src_port, 'UDP')
+
+            if top_listener:
+
+                # queue for data received from the listener
+                listener_q = Queue.Queue()
+                # queue for data received from remote
+                remote_q = Queue.Queue()
+                self.server.logger.debug('Likely listener: %s' % 
+                        top_listener.name)
+                listener_sock = ThreadedUDPClientSocket('localhost', 0, 
+                        listener_q, remote_q, self.server.config, 
+                        self.server.logger, top_listener.port, 
+                        self.server.fwd_table, orig_src_ip, orig_src_port)
+                listener_sock.setDaemon(True)
+                listener_sock.start()
+                remote_sock.setblocking(0)
+
+                # no peek option so process the data already recd
+                remote_q.put(data)
+
+                try:
+                    while True:
+                        readable, writable, exceptional = select.select(
+                                [remote_sock], [], [], .001)
```

Comment:
For the sake of cpu cycles, should we raise the timeout to something like a second?

Reply:
Although I don't think that would violate any timeouts, I think users expect it to be more responsive than 1 second. But users also don't want FakeNet-NG consuming unnecessary CPU. Maybe 100msec (i.e. `0.1`) or so would be a reasonable compromise?

Reply:
A timeout of 1 should still be pretty responsive as it's only controlling how long to wait for a read (or write) from any of the FD's. If the socket is active, the timeout is essentially nil anyways.

Reply:
Ah, you're right, I forgot this was a `select` call.